Data/ByteString/Char8.hs

```diff
@@ -991,34 +992,78 @@ unwords = intercalate (singleton ' ')
 -- ---------------------------------------------------------------------
 -- Reading from ByteStrings
 
--- | readInt reads an Int from the beginning of the ByteString.  If there is no
--- integer at the beginning of the string, it returns Nothing, otherwise
--- it just returns the int read, and the rest of the string.
+-- | Try to read an 'Int' value from the 'ByteString', returning @Just (val,
+-- str)@ on success, where @val@ is the value read and @str@ is the rest of the
+-- input string.  If the sequence of digits decodes to a value larger than can
+-- be represented by an 'Int', the returned value will be 'Nothing'.
+--
+-- 'readInt' does not ignore leading whitespace, the value must start
+-- immediately at the beginning of the input stream.
+--
+-- ==== __Example__
+-- >>> case readInt str of
+-- >>>     Just (n, rest)  -> print n >> gladly rest
+-- >>>     Nothing         -> sadly Nothing
 --
--- Note: This function will overflow the Int for large integers.
 readInt :: ByteString -> Maybe (Int, ByteString)
-readInt as
-    | null as   = Nothing
-    | otherwise =
-        case unsafeHead as of
-            '-' -> loop True  0 0 (B.unsafeTail as)
-            '+' -> loop False 0 0 (B.unsafeTail as)
-            _   -> loop False 0 0 as
-
-    where loop :: Bool -> Int -> Int -> ByteString -> Maybe (Int, ByteString)
-          loop neg !i !n !ps
-              | null ps   = end neg i n ps
-              | otherwise =
-                  case B.unsafeHead ps of
-                    w | w >= 0x30
-                     && w <= 0x39 -> loop neg (i+1)
-                                          (n * 10 + (fromIntegral w - 0x30))
-                                          (B.unsafeTail ps)
-                      | otherwise -> end neg i n ps
-
-          end _    0 _ _  = Nothing
-          end True _ n ps = Just (negate n, ps)
-          end _    _ n ps = Just (n, ps)
+{-# INLINABLE readInt #-}
+readInt bs = case B.uncons bs of
+    Just (w, rest) | w - 0x30 <= 9 -> readDec True bs
+                   | w == 0x2b     -> readDec True rest
+                   | w == 0x2d     -> readDec False rest
+    _                              -> Nothing
```

Comment:
I wish this part was more readable. Replacing characters like `-` with the corresponding hex codes makes this much harder to understand.

The purpose of the first case (`w - 0x30 <= 9`) seems particularly non-obvious. I assume it's an optimization, so we can return `Nothing` early? I think this might be worth a comment.

Reply:
I can add comments, yes, this code is optimised for input starting with a digits, so that's the first case tested, and then perhaps a +/- sign.  Because word8 arithmetic is unsigned, only '0'..'9' satisfy `w - 0x30 <= 9`.

Reply:
I updated the examples and added comments for the optimised branches.